File: docs/scripts/PositionState.js
import CellState from "./CellState.js";
import Player from "./Player.js";
import RotateState from "./RotateState.js";
import State from "./State.js";

export default class PositionState extends State {
    constructor(obj) {
        super(obj);
    }
    position(cell, player) {
        if (this.game.getTurn() !== player) {
            throw new Error("It's not your turn.");
        }
        let { x, y } = cell;
        if (this.game.board[x][y].player !== CellState.EMPTY) {
            throw new Error("Cell must be empty.");
        }
        this.game.positionedPiece = cell;
        this.game.board[x][y].player = player === Player.PLAYER1 ? CellState.PLAYER1 : CellState.PLAYER2;
        this.game.change(new RotateState(this.game));
        return this.game.endOfGame();
    }
    rotate() {
        throw new Error("You can't rotate a piece now.");
    }
    move() {
        throw new Error("You can't move a piece now.");
    }
    changeTurn() {
        throw new Error("You must position a piece in the board.");
    }
}

File: src/main/java/com/chungtoi/PositionState.java
package com.chungtoi;

public class PositionState extends State {

    public PositionState(ChungToi obj) {
        super(obj);
    }

    @Override
    public Winner position(Cell cell, Player player) {
        if (this.game.getTurn() != player) {
            throw new Error("It's not your turn.");
        }
        int x = cell.getX(), y = cell.getY();
        if (this.game.getBoard()[x][y].getPlayer() != CellState.EMPTY) {
            throw new Error("Cell must be empty.");
        }
        this.game.setPositionedPiece(cell);
        this.game.getBoard()[x][y].setPlayer(player == Player.PLAYER1 ? CellState.PLAYER1 : CellState.PLAYER2);
        this.game.change(new RotateState(this.game));
        return this.game.endOfGame();
    }

    @Override
    public void rotate(Cell cell) {
        throw new Error("You can't rotate a piece now.");
    }

    @Override
    public Winner move(Cell beginCell, Cell endCell) {
        throw new Error("You can't move a piece now.");
    }

    @Override
    public void changeTurn() {
        throw new Error("You must position a piece in the board.");
    }
}
